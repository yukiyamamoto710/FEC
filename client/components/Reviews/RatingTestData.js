const testData1 = {
  characteristics: {
    Comfort: {
      id: 86687,
      value: '2.7500000000000000',
    },
    Quality: {
      id: 86688,
      value: '2.0000000000000000',
    },
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
    Width: {
      id: 86686,
      value: '3.2500000000000000',
    },
  },
  product_id: '25821',
  ratings: {
    1: '1',
    4: '1',
    5: '2',
  },
  recommended: {
    false: '1',
    true: '3',
  },
};
const testData2 = {
  characteristics: {
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
  },
  product_id: '25721',
  ratings: {
  },
  recommended: {
  },
};

export { testData1, testData2 };
