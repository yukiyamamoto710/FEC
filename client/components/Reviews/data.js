const Data = {
  Size: [
    'A size too small',
    '½ a size too small',
    'Perfect',
    '½ a size too big',
    'A size too wide',
  ],
  Width: [
    'Too narrow',
    'Slightly narrow',
    'Perfect',
    'Slightly wide',
    'Too wide',
  ],
  Comfort: [
    'Uncomfortable',
    'Slightly uncomfortable',
    'Ok',
    'Comfortable',
    'Perfect',
  ],
  Quality: [
    'Poor',
    'Below average',
    'What I expected',
    'Pretty great',
    'Perfect',
  ],
  Length: [
    'Runs Short',
    'Runs slightly short',
    'Perfect',
    'Runs slightly long',
    'Runs long',
  ],
  Fit: [
    'Runs tight',
    'Runs slightly tight',
    'Perfect',
    'Runs slightly long',
    'Runs long',
  ],
  Stars: [
    'Poor',
    'Fair',
    'Average',
    'Good',
    'Great',
  ],
};

const msg = [
  'For privacy reasons,do not use your full name or email address',
  'For authentication reasons, you will not be emailed',
];

const textExample = [
  'Example: jackson11!',
  'jackson11@email.com',
  'Example: Best purchase ever!',
  'Why did you like the product or not? and need a counter min 50 max 1000',
];

export { Data, msg, textExample };
