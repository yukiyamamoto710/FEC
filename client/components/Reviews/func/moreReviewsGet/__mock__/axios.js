const get = () => Promise.resolve({ data: {data: {results:[]} }});

export default get;
