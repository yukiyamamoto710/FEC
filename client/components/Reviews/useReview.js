import { useEffect, useState } from 'react';

export default function useReview(id) {
  const [countStars, setCountStars] = useState(5);
  const [listUserReview, setListUserReview] = useState([]);
  const [listReported, setListReported] = useState([]);

  useEffect(async () => {
    setCountStars(5);
  }, [id]);

  const starClicked = (num) => {
    setCountStars(num);
  };

  const addUserReview = (obj) => {
    setListUserReview([...listUserReview, obj]);
  };

  const addListReported = (reportId) => {
    setListReported([...listReported, reportId]);
  };

  return ({
    countStars,
    listUserReview,
    listReported,
    starClicked,
    addUserReview,
    addListReported,
  });
}
