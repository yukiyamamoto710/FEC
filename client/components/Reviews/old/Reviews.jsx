import React from 'react';
import axios from 'axios';
import Rbase from './rbase.jsx';
import Rating from './rating.jsx';
// import Popout from './popout.jsx';
// import Warning from './Warning/Warning';
import PopOut from './PopOut/PopOut';

class Reviews extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      productId: '',
      reviewsList: [],
      productRating: '',
      add: false,
      newReview: {
        'Stars': 1,
        'recommend': true,
        'Size': 1,
        'Width': 1,
        'Comfort': 1,
        'Quality': 1,
        'Length': 1,
        'Fit': 1,
        'Title': '',
        'Description': '',
        'Photo': [],
        'Name': '',
        'Email': '',
        'date': '',
        'helpfulness': 0,
        'response': null,
      },
      rvGet: false,
      rtGet: false,
      moreBTNshowed: true,
      rtStarCk: false,
      tempList: [],
      warningItems: [],
    };
    this.reviewsGET = this.reviewsGET.bind(this);
    this.ratingGET = this.ratingGET.bind(this);
    this.sort = this.sort.bind(this);
    this.helpful = this.helpful.bind(this);
    this.notHelpful = this.notHelpful.bind(this);
    this.report = this.report.bind(this);
    this.add = this.add.bind(this);
    this.moreReview = this.moreReview.bind(this);
    this.getTarget = this.getTarget.bind(this);
    this.addReview = this.addReview.bind(this);
    this.loading = this.loading.bind(this);
    this.ratingstar = this.ratingstar.bind(this);
    this.msgClick = this.msgClick.bind(this);
    this.cancel = this.cancel.bind(this);
    this.starCK = this.starCK.bind(this);
    this.cancelWarning = this.cancelWarning.bind(this);
  };

  componentDidMount() {
    let targetId = this.props.id;
    this.setState({
      productId: targetId,
    });
    this.ratingGET( 'reviews/meta', targetId );
    this.reviewsGET( `reviews`, targetId, 2, 'relevant' );
  };

  componentDidUpdate(prevProps) {
    if ( prevProps.id !== this.props.id ) {
      let targetId = this.props.id;
      this.setState({
        productId : targetId,
        rvGet: false,
        rtGet: false,
      });
    this.ratingGET( 'reviews/meta', targetId );
    this.reviewsGET( 'reviews', targetId, 2, 'relevant' );
    };
  };

  getTarget(event) {
    let key = event.target.id;
    let value = event.target.value;
    let HTML = event.target.innerHTML;
    let obj = { ...this.state.newReview };
    if ( key === 'recommend' ) {
      if ( HTML === 'YES' ) {
        value = true;
      } else {
        value = false;
      };
    }

    if ( key === 'Photo' ) {
      if(value === undefined){
        value = ''
      }
      obj[key].push({ url: String(value) });
    } else {
       obj[key] = value;
    }
    this.setState({
      newReview: obj,
    });
  };

  reviewsGET(string, id, count, sort) {
    axios.get( '/get', {
      params: {
        endpoint: `${ string }/?product_id=${ id }&count=${ count }&sort=${ sort }`
      }})
      .then( res =>{
        let arr =  res.data.results;
        const { reviewsList } = this.state;
        let show;
        if ( JSON.stringify(arr) === JSON.stringify(reviewsList) ) {
          show = false;
        } else {
          show = true;
        }
        this.setState({
          reviewsList: arr,
          rvGet: true,
          moreBTNshowed: show,
        });
      })
      .catch( err => console.log);
  };

  ratingGET(string, id) {
    axios.get( '/get', {
      params: {
        endpoint: `${ string }/?product_id=${ id }`
      }})
      .then( res =>{
        this.setState({
          productRating: res.data,
          rtGet: true,
        });
      })
      .catch( err => console.log );
  };

  sort(target) {
    let id = this.state.productId;
    let num = this.state.reviewsList.length;
    this.reviewsGET( 'reviews', id, num, target )
  };

  helpful(target) {
    let arr = this.state.reviewsList.slice();
    if ( arr[ target ][ 'help' ] !== true ) {
      arr[ target ].helpfulness++;
      arr[ target ][ 'help' ] = true;
    };
    //shoudl limit report time with user system
    //should have a put req
    //but not databse to change.
    //use arr[target].proudce_id and PUT /reviews/:review_id/helpful
    this.setState({
      reviewsList: arr,
    });
  };

  notHelpful(target) {
    let arr = this.state.reviewsList.slice();
    if ( arr[ target ][ 'help' ] !== true ) {
      // need to find out which key for not helpful;
      //arr[target].helpfulness++;
      arr[ target ][ 'help' ] = true;
    };
    //shoudl limit report time with user system
    //should have a put req
    //but not databse to change.
    //use arr[target].proudce_id and PUT /reviews/:review_id/helpful
    this.setState({
      reviewsList: arr,
    });
  };

  report(target) {
    let arr = this.state.reviewsList.slice();
    arr.splice( target, 1 );
    this.setState({
      reviewsList: arr,
    });
    //should have a put req
    //but no databse to change.
    //use arr[target].proudce_id and PUT /reviews/:review_id/helpful
    //temple change to State take out first to test function working or not
  };

  add() {
    this.setState({
      add: true,
    });
  };

  moreReview() {
    let oldlen = this.state.reviewsList.length;
    let num = oldlen + 2;
    let targetId = this.state.productId;
    this.reviewsGET( 'reviews', targetId, num, 'relevant' );
  };



  starCK(event) {
    let obj = { ...this.state.newReview };
    let key = event.target.name;
    let value = event.target.id;
    obj[key] = Number(obj[key]) + Number(value);
    this.setState({
      newReview: obj,
    })
  }
  addReview(obj){
  //   if (array.length === 0) {
  //     let obj = { ...this.state.newReview };
      let a = new Date();
      let b = a.toISOString()
      obj.date = b;
  //     obj.body = obj.Description;
  //     obj.photos = obj.Photo;
  //     obj.rating = obj.Stars;
      obj.response = null;
      obj.reviewer_name = obj.name;
  //     obj.summary = obj.Title;
      let arr = [ ...this.state.reviewsList ];
      arr.unshift( obj );
      console.log(arr,)
      this.setState({
        reviewsList: arr,
  //       newReview: {
  //         'Stars': 1,
  //         'recommend': true,
  //         'Size': 1,
  //         'Width': 1,
  //         'Comfort': 1,
  //         'Quality': 1,
  //         'Length': 1,
  //         'Fit': 1,
  //         'Title': '',
  //         'Description': '',
  //         'Photo': [],
  //         'Name': '',
  //         'Email': '',
  //         'date': '',
  //         'helpfulness': 0,
  //         'response': null,
        // },
      //   add:false,
      });
  //   } else {
  //     this.setState({
  //       warningItems: array,
  //     })
  //   }
  };

  addReview(obj) {
    console.log(obj);
    console.log(this.state.reviewsList);
  }

  cancel() {
    this.setState({
      add: false,
    })
  };

  cancelWarning() {
    this.setState({
      warningItems: [],
    });
  }

  loading() {
    const { rvGet, rtGet } = this.state;
    if ( rvGet === true && rtGet === true ) {
      const { reviewsList, productRating, moreBTNshowed, newReview, add, warningItems } = this.state;
      return (
        <div
          style = { bas }>
          <div>RATINGS REVIEWS</div>
          <div
            style = { base }>
              {/* { warningItems.length !== 0?
                <Warning
                  warningItems = { warningItems }
                  show = { this.cancelWarning }
                />
              :
                null } */}

              { add === true?
              <PopOut
                  addReview = { this.addReview }
                  // review = { newReview }
                  // getTarget = { this.getTarget }
                  // starCK = { this.starCK }
                  cancel = {this.cancel}
                  data = {this.state.productRating}
                />
              :
                null }
            <Rating
              rating = { productRating }
              ratingstar = { this.ratingstar }/>
            <Rbase
              list = { reviewsList }
              sort = { this.sort }
              helpful ={ this.helpful }
              notHelpful ={ this.notHelpful }
              report ={ this.report }
              moreReview = { this.moreReview }
              addfunc = { this.add }
              msgClick = { this.msgClick }
              moreBTN = { moreBTNshowed }/>
          </div>
        </div>
      );
    } else {
      return <div>Loading...</div>
    };
  };

  ratingstar(num) {
    const { reviewsList, tempList, rtStarCk } = this.state;
    let arr;
    if ( rtStarCk === false ) {
      arr = [ ...reviewsList ];
    } else {
      arr = [ ... tempList ];
    }
    let arr1 = arr.filter( i => i.rating <= Number( num ));
    if ( JSON.stringify( arr1 ) === JSON.stringify( reviewsList )) {
      this.setState({
        reviewsList: arr,
        tempList:[],
        rtStarCk: false,
      })
    } else {
      this.setState({
        rtStarCk: true,
        reviewsList: arr1,
        tempList : arr,
      })
    }
  };

  msgClick(event) {
    let arr = [ ...this.state.reviewsList ];
    let index = event.target.id;
    if (arr[index]['click'] === undefined) {
      arr[index]['click'] = true;
    } else {
      arr[index]['click'] = undefined;
    }
    this.setState({
      reviewsList: arr,
    })
  }

  render() {
    return (
      <div>
        { this.loading() }
      </div>
    );
  };
};

export default Reviews;

const base = {
  display: 'flex',
  justifyContent:'center',
}

const bas= {
  justifyContent: 'center',
}