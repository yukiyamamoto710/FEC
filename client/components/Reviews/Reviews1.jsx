import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Rating from './Rating/Rating';
import ReviewListBase from './ReviewListBase/ReviewListBase';
import useReview from './useReview';

export default function Reviews(props) {
  const { id, productRating } = props;
  const {
    countStars,
    listUserReview,
    listReported,
    starClicked,
    addUserReview,
    addListReported,
  } = useReview(id);

  return (
    <div>
      <h3>Reviews & Rating</h3>
      <div
        data-testid="reviewContainer"
        className="reviewContainer"
      >
        <Rating
          rating={productRating}
          starClicked={starClicked}
        />
        <ReviewListBase
          rating={productRating}
          id={id}
          listReported={listReported}
          addListReported={addListReported}
          listUserReview={listUserReview}
          addUserReview={addUserReview}
          stars={countStars}
        />
      </div>
    </div>
  );
}

Reviews.propTypes = {
  productRating: PropTypes.shape({
    characteristics: PropTypes.shape({}),
    product_id: PropTypes.string,
    ratings: PropTypes.shape({}),
    recommended: PropTypes.shape({}),
  }),
  id: PropTypes.number,
};

Reviews.defaultProps = {
  productRating: {
    characteristics: {},
    product_id: "25711",
    ratings: {},
    recommended: {},
  },
  id: 25711,
};
