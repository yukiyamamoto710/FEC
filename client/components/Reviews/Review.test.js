/**
 * @jest-environment jsdom
 */
import React from 'react';
import axios from 'axios';
import renderer from 'react-test-renderer';
import {
  render,
  screen,
  cleanup,
  waitFor,
  waitForElement,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Reviews from './Reviews1';
import testData1 from './RatingTestData'

jest.mock('axios');

afterEach(cleanup);
it('isRatingLoad false', () => {
  const testId = 25711;
  const { getByTestId } = render(<Reviews id={testId} productRating={testData1}/>);
  expect(getByTestId('reviewContainer')).toBeInTheDocument();
});

it('matches snapShot', () => {
  const tree = renderer.create(
    <Reviews
      id={25711}
      productRating={testData1}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

