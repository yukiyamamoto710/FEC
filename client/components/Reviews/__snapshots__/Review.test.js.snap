// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<div>
  <h3>
    Reviews & Rating
  </h3>
  <div
    className="reviewContainer"
    data-testid="reviewContainer"
  >
    <div
      className="rating"
      data-testid="rating"
    >
      <div
        className="ratingHead"
        data-testid="head"
      >
        0
        <div
          className="ratingHead"
          data-testid="stars"
        >
          <div>
            <div
              className="ratingStar emptyStar"
              data-testid="button0"
              id=" 1"
              label="emptyStar"
              name=""
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            />
          </div>
          <div>
            <div
              className="ratingStar emptyStar"
              data-testid="button1"
              id=" 2"
              label="emptyStar"
              name=""
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            />
          </div>
          <div>
            <div
              className="ratingStar emptyStar"
              data-testid="button2"
              id=" 3"
              label="emptyStar"
              name=""
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            />
          </div>
          <div>
            <div
              className="ratingStar emptyStar"
              data-testid="button3"
              id=" 4"
              label="emptyStar"
              name=""
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            />
          </div>
          <div>
            <div
              className="ratingStar emptyStar"
              data-testid="button4"
              id=" 5"
              label="emptyStar"
              name=""
              onClick={[Function]}
              onKeyPress={[Function]}
              role="button"
              tabIndex={0}
            />
          </div>
        </div>
      </div>
      <h5
        data-testid="text"
      >
        0% of reviews recommend this product
      </h5>
      <div
        data-testid="starListR"
      >
        <div
          className="starListContainer"
          data-testid="starList"
        >
          <div>
            <div
              className="starListBase"
            >
              <button
                className="starListText"
                data-testid="startListButton5"
                id="5"
                onClick={[Function]}
                type="button"
              >
                5 Stars
              </button>
              <div
                className="starRatingBarBase"
                data-testid="starRatingBar"
              >
                <div
                  style={
                    Object {
                      "backgroundColor": "rgb(102, 162, 182)",
                      "border": "5px solid rgb(102, 162, 182)",
                      "width": "NaN%",
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <div
              className="starListBase"
            >
              <button
                className="starListText"
                data-testid="startListButton4"
                id="4"
                onClick={[Function]}
                type="button"
              >
                4 Stars
              </button>
              <div
                className="starRatingBarBase"
                data-testid="starRatingBar"
              >
                <div
                  style={
                    Object {
                      "backgroundColor": "rgb(102, 162, 182)",
                      "border": "5px solid rgb(102, 162, 182)",
                      "width": "NaN%",
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <div
              className="starListBase"
            >
              <button
                className="starListText"
                data-testid="startListButton3"
                id="3"
                onClick={[Function]}
                type="button"
              >
                3 Stars
              </button>
              <div
                className="starRatingBarBase"
                data-testid="starRatingBar"
              >
                <div
                  style={
                    Object {
                      "backgroundColor": "rgb(102, 162, 182)",
                      "border": "5px solid rgb(102, 162, 182)",
                      "width": "NaN%",
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <div
              className="starListBase"
            >
              <button
                className="starListText"
                data-testid="startListButton2"
                id="2"
                onClick={[Function]}
                type="button"
              >
                2 Stars
              </button>
              <div
                className="starRatingBarBase"
                data-testid="starRatingBar"
              >
                <div
                  style={
                    Object {
                      "backgroundColor": "rgb(102, 162, 182)",
                      "border": "5px solid rgb(102, 162, 182)",
                      "width": "NaN%",
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div>
            <div
              className="starListBase"
            >
              <button
                className="starListText"
                data-testid="startListButton1"
                id="1"
                onClick={[Function]}
                type="button"
              >
                1 Stars
              </button>
              <div
                className="starRatingBarBase"
                data-testid="starRatingBar"
              >
                <div
                  style={
                    Object {
                      "backgroundColor": "rgb(102, 162, 182)",
                      "border": "5px solid rgb(102, 162, 182)",
                      "width": "NaN%",
                    }
                  }
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      loading...
    </div>
  </div>
</div>
`;
